File: intellij-lsp/src/com/github/gtache/lsp/settings/server/parser/ConfigurationParser.scala
package com.github.gtache.lsp.settings.server.parser

import java.io.{File, FileReader}

import com.github.gtache.lsp.settings.server.LSPConfiguration
import com.google.gson
import com.google.gson._
import com.intellij.openapi.util.io.FileUtilRt
import javax.xml.parsers.DocumentBuilderFactory

import scala.collection.JavaConverters._
import scala.collection.mutable

object ConfigurationParser {

  def getConfiguration(file: File): LSPConfiguration = {
    val parser = forFile(file)
    parser.map(p => {
      p.parse(file)
    }).orNull
  }

  def getConfiguration(doc: String, typ: ConfigType): LSPConfiguration = {
    val parser = forType(typ)
    val file = FileUtilRt.createTempFile("config", "." + ConfigType.toExt(typ), true)
    parser.map(p => p.parse(file)).orNull
  }

  def forType(typ: ConfigType): Option[ConfigurationParser] = {
    typ match {
      case ConfigType.FLAT => Some(new FlatParser())
      case ConfigType.JSON => Some(new JsonParser())
      case ConfigType.XML => Some(new XmlParser())
      case _ => None
    }
  }

  def forFile(file: File): Option[ConfigurationParser] = {
    val name = file.getName
    val idx = name.lastIndexOf('.')
    if (idx + 1 < name.length) {
      forExt(name.drop(idx + 1))
    } else None
  }

  def forExt(ext: String): Option[ConfigurationParser] = {
    forType(ConfigType.forExt(ext))
  }

  def combineConfigurations(firstConfig: Map[String, Map[String, AnyRef]], secondConfig: Map[String, Map[String, AnyRef]]): Map[String, Map[String, AnyRef]] = {
    val concatMap = mutable.Map[String, mutable.Map[String, AnyRef]]()
    firstConfig.keySet.foreach(key => {
      concatMap.update(key, mutable.Map(firstConfig(key).toSeq: _*))
    })
    secondConfig.keySet.foreach(key => {
      if (concatMap.contains(key)) {
        secondConfig(key).keySet.foreach(subkey => concatMap(key).update(subkey, secondConfig(key)(subkey)))
      } else {
        concatMap.update(key, mutable.Map(secondConfig(key).toSeq: _*))
      }
    })
    concatMap.map(pair => pair._1 -> pair._2.toMap).toMap
  }
}

trait ConfigurationParser {
  def parse(file: File): LSPConfiguration
}

class JsonParser extends ConfigurationParser {
  override def parse(file: File): LSPConfiguration = {
    val reader = new FileReader(file)
    val jsonObject = new gson.JsonParser().parse(reader).getAsJsonObject
    val configMap = Map[String, Map[String, AnyRef]]("global" -> Map[String, AnyRef]())

    def flatten(scope: String, key: String, elem: JsonElement, map: Map[String, Map[String, AnyRef]]): Map[String, Map[String, AnyRef]] = {
      var trueScope = scope
      var trueKey = key
      if (scope == null) {
        if (trueKey == null) {
          trueScope = null
          trueKey = null
        }
        else if (trueKey != null && trueKey.startsWith("[") && trueKey.endsWith("]")) {
          if (!elem.isJsonObject) {
            throw new IllegalArgumentException("Check JSON file " + file.getAbsolutePath)
          }
          trueScope = key
          trueKey = null
        } else {
          trueScope = "global"
          trueKey = key
        }
      }
      val updatedMap = if (trueScope != null && !map.contains(trueScope)) {
        map.updated(trueScope, Map[String, AnyRef]())
      } else map
      elem match {
        case obj: JsonObject =>
          obj.keySet().asScala.map(key => {
            flatten(trueScope, if (trueKey != null) trueKey + "." + key else key, obj.get(key), updatedMap)
          }).fold(updatedMap)((map1, map2) => {
            ConfigurationParser.combineConfigurations(map1, map2)
          })
        case arr: JsonArray =>
          val javaArr = arr.asScala.toArray
          updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, javaArr))
        case _: JsonNull =>
          updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, null))
        case prim: JsonPrimitive =>
          if (prim.isBoolean) {
            updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, prim.getAsBoolean.asInstanceOf[java.lang.Boolean]))
          } else if (prim.isNumber) {
            updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, prim.getAsNumber))
          } else if (prim.isString) {
            updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, prim.getAsString))
          } else {
            updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, null))
          }
        case _ => updatedMap.updated(trueScope, updatedMap(trueScope).updated(trueKey, null))
      }
    }

    LSPConfiguration(flatten(null, null, jsonObject, configMap))
  }
}

class XmlParser extends ConfigurationParser {
  override def parse(file: File): LSPConfiguration = {
    val factory = DocumentBuilderFactory.newInstance()
    val builder = factory.newDocumentBuilder()
    val doc = builder.parse(file)
    val root = doc.getDocumentElement
    throw new UnsupportedOperationException
  }
}

class FlatParser extends ConfigurationParser {
  override def parse(file: File): LSPConfiguration = {
    throw new UnsupportedOperationException
  }
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/settings/server/parser/ConfigurationParser.kt
package com.github.gtache.lsp.settings.server.parser

import com.github.gtache.lsp.settings.server.Configuration
import com.google.gson.*
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.io.FileUtilRt
import java.io.File
import java.io.FileReader
import javax.xml.parsers.DocumentBuilderFactory


interface ConfigurationParser {
    fun parse(file: File): Configuration

    companion object {
        fun getConfiguration(file: File): Configuration? {
            val parser = forFile(file)
            return parser?.parse(file)
        }

        fun getConfiguration(doc: String, typ: ConfigType): Configuration? {
            val parser = forType(typ)
            val file = FileUtilRt.createTempFile("config", "." + ConfigType.toExt(typ), true)
            return parser?.parse(file)
        }


        private fun forType(typ: ConfigType?): ConfigurationParser? {
            return when (typ) {
                ConfigType.FLAT -> FlatParser()
                ConfigType.JSON -> JsonParser()
                ConfigType.XML -> XmlParser()
                else -> null
            }
        }

        private fun forFile(file: File): ConfigurationParser? {
            val name = file.name
            val idx = name.lastIndexOf('.')
            return if (idx + 1 < name.length) {
                forExt(name.drop(idx + 1))
            } else null
        }

        private fun forExt(ext: String): ConfigurationParser? {
            return forType(ConfigType.fromExt(ext))
        }

        fun combineConfigurations(firstConfig: Map<String, Map<String, Any?>>, secondConfig: Map<String, Map<String, Any?>>): Map<String, Map<String, Any?>> {
            val concatMap = HashMap<String, MutableMap<String, Any?>>()
            firstConfig.keys.forEach { key ->
                concatMap[key] = HashMap(firstConfig[key])
            }
            secondConfig.keys.forEach { key ->
                if (concatMap.contains(key)) {
                    secondConfig[key]!!.keys.forEach { subkey -> concatMap[key]!![subkey] = secondConfig[key]!![subkey]!! }
                } else {
                    concatMap[key] = HashMap(secondConfig[key])
                }
            }
            return concatMap.map { pair -> pair.key to pair.value.toMap() }.toMap()
        }
    }

    class JsonParser : ConfigurationParser {

        companion object {
            private val logger: Logger = Logger.getInstance(JsonParser::class.java)
        }

        override fun parse(file: File): Configuration {
            val reader = FileReader(file)
            if (file.length() == 0L) {
                return Configuration.emptyConfiguration
            } else {
                try {
                    val json = com.google.gson.JsonParser.parseReader(reader)
                    if (json != null && json.isJsonObject) {
                        val jsonObject = json.asJsonObject
                        val configMap = mapOf("global" to HashMap<String, Any>())

                        fun flatten(scope: String?, key: String?, elem: JsonElement, map: Map<String, Map<String, Any?>>): Map<String, Map<String, Any?>> {
                            val trueScope: String
                            val trueKey: String
                            if (scope == null) {
                                if (key == null) {
                                    trueScope = "global"
                                    trueKey = ""
                                } else if (key.startsWith("<") && key.endsWith(">")) {
                                    if (!elem.isJsonObject) {
                                        throw IllegalArgumentException("Check JSON file ${file.absolutePath}")
                                    }
                                    trueScope = key
                                    trueKey = ""
                                } else {
                                    trueScope = "global"
                                    trueKey = key
                                }
                            } else {
                                trueScope = scope
                                trueKey = key ?: ""
                            }
                            val updatedMap = (if (!map.contains(trueScope)) {
                                map + (trueScope to HashMap())
                            } else map).toMutableMap()
                            val subMap = updatedMap[trueScope]!!
                            when (elem) {
                                is JsonObject ->
                                    return elem.keySet().map { k ->
                                        flatten(trueScope, if (trueKey.isNotEmpty()) "$trueKey.$k" else k, elem.get(k), updatedMap)
                                    }.fold(updatedMap) { map1, map2 -> combineConfigurations(map1, map2).toMutableMap() }
                                is JsonArray -> {
                                    return updatedMap + (trueScope to subMap + (trueKey to elem))
                                }
                                is JsonNull -> return updatedMap + (trueScope to subMap + (trueKey to null))
                                is JsonPrimitive ->
                                    return if (elem.isBoolean) {
                                        updatedMap + (trueScope to subMap + (trueKey to elem.asBoolean))
                                    } else if (elem.isNumber) {
                                        updatedMap + (trueScope to subMap + (trueKey to elem.asNumber))
                                    } else if (elem.isString) {
                                        updatedMap + (trueScope to subMap + (trueKey to elem.asString))
                                    } else {
                                        updatedMap + (trueScope to subMap + (trueKey to null))
                                    }
                                else -> return updatedMap + (trueScope to subMap + (trueKey to null))
                            }
                        }
                        return Configuration(flatten("global", null, jsonObject, configMap))
                    } else {
                        return Configuration.invalidConfiguration
                    }
                } catch (t: Throwable) {
                    logger.warn("Error parsing JSON", t)
                    return Configuration.invalidConfiguration
                }
            }
        }
    }

    class XmlParser : ConfigurationParser {
        override fun parse(file: File): Configuration {
            val factory = DocumentBuilderFactory.newInstance()
            val builder = factory.newDocumentBuilder()
            val doc = builder.parse(file)
            val root = doc.documentElement
            throw UnsupportedOperationException()
        }
    }

    class FlatParser : ConfigurationParser {
        override fun parse(file: File): Configuration {
            throw UnsupportedOperationException()
        }
    }
}